Added support for groups and group milestones.

ChrisCalculus opened the pull request:
This PR adds support for Gitlab Groups and Group milestones to the tap. When syncing groups, one can choose between syncing all projects contained in the group, or make a sub-selection of projects.

cmerrick commented:
Hi @ChrisCalculus, thanks for your contribution!

In order for us to evaluate and accept your PR, we ask that you [sign a contribution license agreement](https://stitch-cla-enforcer.herokuapp.com/). It's all electronic and will take just minutes.

cmerrick commented:
You did it @ChrisCalculus!

Thank you for signing the Singer Contribution License Agreement.

KAllan357 commented on tap_gitlab/__init__.py:
```diff
@@ -185,6 +195,29 @@ def sync_users(project):
+    data = request(url).json()
+    time_extracted = utils.now()
+
+    with Transformer(pre_hook=format_timestamp) as transformer:
```
Its a good practice to limit the scope of the Transformer for as long as its needed. In this case, I'd like to see the code after the transformation be unindented so the Transformer is used and discarded as the tap moves forward into the other for loop.

ChrisCalculus replied:
I did so, and also applied it to sync_project. Please check if this what you expected.

KAllan357 commented on tap_gitlab/schemas/groups.json:
```diff
@@ -0,0 +1,59 @@
+        "full_path": {
+            "type": ["null", "string"]
+        },
+        "id": {
```
This appears to be duplicating / shadowing the `id` element higher up on line 4.

ChrisCalculus replied:
My bad, I've corrected it.

KAllan357 commented on tap_gitlab/__init__.py:
```diff
@@ -218,27 +251,35 @@ def sync_project(pid):
+    for gid in gids:
+        sync_group(gid, pids)
+
+    if not gids:
```
Would you ever want to sync both groups and projects? It should be safe to not guard this for loop, right?

ChrisCalculus replied:
Yes, please see my detailed explanation in the comment below or the README file.

ChrisCalculus commented:
Hi @KAllan357,

thanks for your feedback. I've provided fixes as you suggested. Please see my new commit.

In detail:

> Could you please provide a bit of a description in this pull request. Just a general summary of what's changing / being added would be perfect.

Done.

> Could you add some text to the README.md to indicate that we'll also be syncing groups and group milestones?

Done.

> If I understand correctly, this change will require a new field in the Stitch UI. It looks like it wants to perform the same was as our current project field - a space separated list of GitLab groups. Is that correct?

TL;DR: Yes.

As groups can 'contain' projects and the current tap allows for a selection of projects, it seemed logical to me that you'd also want to configure a selection of groups. 
A user might want to sync all projects contained in a group or only sync a sub-set of projects.
I've implemented it in such a way that 3 scenarios are possible:
* User only configures groups in the config.json (and leaves projects empty): all projects contained in these groups will be synced.
* User only configures projects in the config.json (and leaves groups empty): only selected projects will be synced.
* User configures groups and projects in the config.json: only selected projects (if) part of selected groups will be synced.

Note: Unfortunately groups don't provide a date field to be tracked as part of the state.

I've added responses to your individual comments.

KAllan357 commented:
Thanks for the detailed responses. This looks good now. :ship:

KAllan357 commented:
I'm going to merge this now and release it over the next few days. We'll need a little bit of time to add the new UI element.

ChrisCalculus commented:
Great, thanks @KAllan357
